// zklocus/build/src/blockchain/contracts/tokens/zkl/ZKLContract.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { SmartContract, state, State, method, Permissions, UInt64, PublicKey, Signature, } from 'o1js';
const tokenSymbol = 'ZKL';
/**
 * Token contract for the $ZKL token - the ecosystem token for zkLocus.
 *
 * The token is being developed iteratively, including it's interface.
 *
 */
export class ZKLContract extends SmartContract {
    constructor() {
        super(...arguments);
        this.SUPPLY_CAP = UInt64.from(1n ** 18n);
        this.circulatingSupply = State();
    }
    deploy(args) {
        super.deploy(args);
        const permissionToEdit = Permissions.proof();
        this.account.permissions.set({
            ...Permissions.default(),
            editState: permissionToEdit,
            setTokenSymbol: permissionToEdit,
            send: permissionToEdit,
            receive: permissionToEdit,
        });
    }
    init() {
        super.init();
        this.account.tokenSymbol.set(tokenSymbol);
        this.circulatingSupply.set(UInt64.zero);
    }
    mint(receiverAddress, amount, adminSignature) {
        let totalCirculatingSupply = this.circulatingSupply.get();
        this.circulatingSupply.requireEquals(totalCirculatingSupply);
        const newTotalAmountInCirculation = totalCirculatingSupply.add(amount);
        newTotalAmountInCirculation.assertLessThanOrEqual(this.SUPPLY_CAP, "Cannot mint above the supply cap.");
        adminSignature
            .verify(this.address, amount.toFields().concat(receiverAddress.toFields()))
            .assertTrue();
        this.token.mint({
            address: receiverAddress,
            amount,
        });
        this.circulatingSupply.set(newTotalAmountInCirculation);
    }
}
__decorate([
    state(UInt64),
    __metadata("design:type", Object)
], ZKLContract.prototype, "circulatingSupply", void 0);
__decorate([
    method,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", []),
    __metadata("design:returntype", void 0)
], ZKLContract.prototype, "init", null);
__decorate([
    method,
    __metadata("design:type", Function),
    __metadata("design:paramtypes", [PublicKey,
        UInt64,
        Signature]),
    __metadata("design:returntype", void 0)
], ZKLContract.prototype, "mint", null);
