// zklocus/build/src/api/sha3/SHA3.js
import { Bytes, Hash, Poseidon } from "o1js";
/**
 * Represents a 64-byte array.
 */
export class Bytes64 extends Bytes(64) {
}
/**
 * Represents an SHA3-512 hash fully compatible with Zero-Knowledge circuits.
 */
export class SHA3_512 {
    /**
     * Creates a new instance of SHA3_512.
     * @param _preimage The preimage string to be hashed.
     */
    constructor(_preimage) {
        this._preimage = _preimage;
        this._preimageBytes = Bytes.fromString(_preimage);
        this._digest = Hash.SHA3_512.hash(this._preimageBytes);
    }
    /**
     * Gets the provable type that can be used in Zero-Knowledge circuits.
     * @returns The provable type
     */
    static get provable() {
        return Bytes64.provable;
    }
    /**
     * Gets the `Bytes64` representation of the preimage.
     * @returns The `Bytes64` representation of the preimage.
     */
    get preimageBytes() {
        return this._preimageBytes;
    }
    /**
     * Gets the `Bytes64` representation of the digest.
     * @returns The `Bytes64` representation of the digest.
     */
    get digest() {
        return this._digest;
    }
    /**
     * Gets raw preimage string. This is the string that was used to create the SHA3_512 instance.
     * @returns The raw preimage string.
     */
    get preimage() {
        return this._preimage;
    }
    /**
     * Returns the Poseidon.hash() of the SHA3-512 digest of the preimage. This is the commitment used
     * in the Zero-Knowledge circuits of zkLocus.
     * @returns The Poseidon.hash() of the SHA3-512 digest of the preimage.
     */
    poseidonHash() {
        return Poseidon.hash(this._digest.toFields());
    }
}
