// zklocus/build/src/api/adopters/ZKThreePointPolygonToThreePointPolygonAdopter.js
/*
    This file defines the Adaptor that converts/adapts a ZKThreePointPolygon into a ThreePointPolygon.

    It's defined in a separate file and exported as a default export, due to TS4094 which presents issues
    when an exported class experssion contains private or protected members. The approach followed in this file
    serves as a workaround for this issue.

    See: https://github.com/microsoft/TypeScript/issues/30355
*/
import { ThreePointPolygon } from "../../model/Geography";
export default function (Base) {
    return class extends Base {
        rawValue() {
            return this.vertices;
        }
        normalizedValue() {
            return [
                this.vertices[0].toZKValue(),
                this.vertices[1].toZKValue(),
                this.vertices[2].toZKValue()
            ];
        }
        toZKValue() {
            const vertices = this.vertices.map(vertex => vertex.toZKValue());
            const threePointPolygon = new ThreePointPolygon({
                vertice1: vertices[0],
                vertice2: vertices[1],
                vertice3: vertices[2]
            });
            return threePointPolygon;
        }
    };
}
