// o1js/dist/node/bindings/mina-transaction/transaction-leaves-bigint.js
import { Field, Bool, UInt32, UInt64, Sign, } from '../../provable/field-bigint.js';
import { PublicKey } from '../../provable/curve-bigint.js';
import { derivedLeafTypesSignable } from './derived-leaves.js';
import { createEvents } from '../../lib/events.js';
import { Poseidon, HashHelpers, packToFields, } from '../../provable/poseidon-bigint.js';
import { mocks } from '../crypto/constants.js';
export { PublicKey, Field, Bool, AuthRequired, UInt64, UInt32, Sign, TokenId };
export { Events, Actions, ZkappUri, TokenSymbol, ActionState, VerificationKeyHash, ReceiptChainHash, StateHash, };
const { TokenId, StateHash, TokenSymbol, AuthRequired, ZkappUri } = derivedLeafTypesSignable({ Field, Bool, HashHelpers, packToFields });
const { Events, Actions } = createEvents({ Field, Poseidon });
const ActionState = {
    ...Field,
    empty: Actions.emptyActionState,
};
const VerificationKeyHash = {
    ...Field,
    empty: () => Field(mocks.dummyVerificationKeyHash),
};
const ReceiptChainHash = {
    ...Field,
    empty: () => HashHelpers.emptyHashWithPrefix('CodaReceiptEmpty'),
};
