// o1js/dist/web/bindings/lib/provable-snarky.js
import { createDerivers, createHashInput, } from './provable-generic.js';
// external API
export { provable, provablePure, provableTuple, provableFromClass, };
// internal API
export { HashInput, };
const HashInput = createHashInput();
const { provable } = createDerivers();
function provablePure(typeObj) {
    return provable(typeObj, { isPure: true });
}
function provableTuple(types) {
    return provable(types);
}
function provableFromClass(Class, typeObj) {
    let raw = provable(typeObj);
    return {
        sizeInFields: raw.sizeInFields,
        toFields: raw.toFields,
        toAuxiliary: raw.toAuxiliary,
        fromFields(fields, aux) {
            return construct(Class, raw.fromFields(fields, aux));
        },
        check(value) {
            if (Class.check !== undefined) {
                Class.check(value);
            }
            else {
                raw.check(value);
            }
        },
        toInput: raw.toInput,
        toJSON: raw.toJSON,
        fromJSON(x) {
            return construct(Class, raw.fromJSON(x));
        },
        empty() {
            return construct(Class, raw.empty());
        },
    };
}
function construct(Class, value) {
    let instance = Object.create(Class.prototype);
    return Object.assign(instance, value);
}
