// zklocus/build/src/api/proofs/middleware/CachingGeoPointInPolygonProofMiddleware.js
/**
 * This is a middleware that adds caching and cache invalidation to the verification of a proof. It should be used as a decorator on a class that extends ZKGeoPointInPolygonProof.
 * @param Base - The base class to extend
 * @returns A class that extends the base class and adds caching to the verification of the proof
 */
export default function (Base) {
    return class extends Base {
        constructor() {
            super(...arguments);
            this.isVerified = false;
        }
        verify() {
            if (!this.isVerified) {
                super.verify();
                this.isVerified = true;
            }
        }
        verifyIf(condition) {
            if (!this.isVerified && condition) {
                super.verify();
                this.isVerified = true;
            }
        }
        get isCached() {
            return this.isVerified;
        }
        setProof(proof) {
            super.setProof(proof);
            this.isVerified = false;
        }
        async AND(other) {
            const result = await super.AND(other);
            this.isVerified = false;
            return result;
        }
    };
}
